File: backend/src/types/entities.py
"""Pydantic mirrors of the 14 BRD §9 entities plus `RebalancingThreshold`.

These models carry the persistence shape from data-models.md §4: field names are
the BRD's, snake_case and unchanged; timestamps and dates are ISO-8601 / YYYY-MM-DD
strings because SQLite stores them as TEXT; money and percentage values are
`Decimal`, never float (NFR-01). `User.password_hash` is present because the table
has it — the never-serialise rule is enforced by `types/responses.py`, not by
omitting the column here.
"""

from decimal import Decimal

from pydantic import BaseModel, ConfigDict

from src.types.enums import (
    AuditAction,
    AuditEntityType,
    RecommendationStatus,
    RiskBand,
    Role,
    TradeAction,
)

AuditDetailValue = str | int | bool


class Entity(BaseModel):
    """Base for the persistence mirrors: closed shape, immutable once built."""

    model_config = ConfigDict(extra="forbid", frozen=True)


class QuestionOption(Entity):
    """One selectable answer inside a questionnaire question."""

    value: str
    label: str
    points: int


class Question(Entity):
    """One question of `RiskBandRule.questionnaire_json`."""

    question_id: str
    text: str
    options: list[QuestionOption]


class Questionnaire(Entity):
    """The `questionnaire_json` document (data-models.md §4.5)."""

    questions: list[Question]


class BandRange(Entity):
    """One closed, integer point range mapping a score to a risk band."""

    risk_band: RiskBand
    min_points: int
    max_points: int


class ScoringRules(Entity):
    """The `scoring_rules_json` document — contiguous, non-overlapping ranges."""

    bands: list[BandRange]


class AllocationEntry(Entity):
    """One asset-class weight in stored form: integer basis points summing to 10000."""

    asset_class_id: int
    percent_bps: int


class AllocationSet(Entity):
    """The `allocations_json` document (data-models.md §4.6)."""

    allocations: list[AllocationEntry]


class ProposedAction(Entity):
    """One BUY/SELL leg of a rebalancing proposal; `drift_percent` may be negative."""

    asset_class_id: int
    asset_class_code: str
    action: TradeAction
    amount: Decimal
    units: Decimal
    drift_percent: Decimal


class ProposedActions(Entity):
    """The `proposed_actions_json` document, with the threshold it was judged against."""

    actions: list[ProposedAction]
    threshold_bps: int
    threshold_version: int


class User(Entity):
    """data-models.md §4.1."""

    id: int
    email: str
    password_hash: str
    role: Role
    created_at: str


class Customer(Entity):
    """data-models.md §4.2."""

    id: int
    user_id: int
    kyc_verified: bool
    created_at: str


class RiskProfileAnswer(Entity):
    """data-models.md §4.3 — append-only."""

    id: int
    customer_id: int
    question_id: str
    answer_value: str
    submitted_at: str


class RiskBandAssignment(Entity):
    """data-models.md §4.4 — append-only; `rule_version` pins the rule used."""

    id: int
    customer_id: int
    risk_band: RiskBand
    rule_version: int
    assigned_at: str


class RiskBandRule(Entity):
    """data-models.md §4.5 — insert-only, immutable once published."""

    id: int
    version: int
    questionnaire_json: Questionnaire
    scoring_rules_json: ScoringRules
    published_at: str
    is_active: bool


class AllocationTemplate(Entity):
    """data-models.md §4.6 — insert-only, immutable once published."""

    id: int
    version: int
    risk_band: RiskBand
    allocations_json: AllocationSet
    published_at: str
    is_active: bool


class Goal(Entity):
    """data-models.md §4.7 — `target_amount` is money, `target_date` a YYYY-MM-DD string."""

    id: int
    customer_id: int
    target_amount: Decimal
    target_date: str
    priority: int
    created_at: str
    updated_at: str


class GoalProgressSnapshot(Entity):
    """data-models.md §4.8 — append-only; `percent_complete` caps at 200.00."""

    id: int
    goal_id: int
    current_value: Decimal
    percent_complete: Decimal
    snapshot_at: str
    price_date: str


class AssetClass(Entity):
    """data-models.md §4.9 — master data with a unique `code`."""

    id: int
    code: str
    name: str


class NavSnapshot(Entity):
    """data-models.md §4.10 — append-only price per asset class per simulated date."""

    id: int
    asset_class_id: int
    price_date: str
    nav_value: Decimal


class Holding(Entity):
    """data-models.md §4.11 — a current position, revalued on each advance-a-day."""

    id: int
    customer_id: int
    asset_class_id: int
    current_value: Decimal
    as_of_date: str


class RebalancingRecommendation(Entity):
    """data-models.md §4.12 — `recommendation_id` is the public UUID used in API paths."""

    id: int
    customer_id: int
    recommendation_id: str
    proposed_actions_json: ProposedActions
    status: RecommendationStatus
    generated_at: str
    resolved_at: str | None = None


class AdvisorOverride(Entity):
    """data-models.md §4.13 — append-only; `reason` mandatory, `note` optional."""

    id: int
    customer_id: int
    advisor_id: int
    previous_band: RiskBand
    new_band: RiskBand
    reason: str
    note: str | None = None
    created_at: str


class AuditLogEntry(Entity):
    """data-models.md §4.14 — insert-only; `details_json` carries ids and enums only."""

    id: int
    entity_type: AuditEntityType
    entity_id: str
    actor_id: int
    actor_role: Role
    action: AuditAction
    timestamp: str
    details_json: dict[str, AuditDetailValue]


class RebalancingThreshold(Entity):
    """data-models.md §4.15 — the one percent-class quantity typed `int` (basis points)."""

    id: int
    version: int
    threshold_bps: int
    published_at: str
    is_active: bool

File: frontend/src/types/entities.ts
/**
 * TypeScript mirrors of the backend Pydantic entities (E1-S1 AC5).
 *
 * Field names are identical to `backend/src/types/entities.py` — snake_case, no key
 * mapping — so a JSON response deserialises straight into these shapes.
 *
 * Money, percentage and unit values are `string`, never `number`: a JSON number
 * becomes an IEEE-754 double in JavaScript, which is what NFR-01 forbids
 * (data-models.md §2). The two exceptions are `threshold_bps` and `percent_bps`,
 * whose canonical representation is an integer count of basis points.
 */

export type Role = 'customer' | 'advisor' | 'admin' | 'compliance';

export type RiskBand = 'CONSERVATIVE' | 'MODERATE' | 'AGGRESSIVE';

export type RecommendationStatus = 'pending' | 'accepted' | 'dismissed';

export type TradeAction = 'BUY' | 'SELL';

export type AuditEntityType =
  | 'RiskBandAssignment'
  | 'AllocationRecommendation'
  | 'RebalancingRecommendation'
  | 'AdvisorOverride'
  | 'ManualRecommendation'
  | 'RiskBandRule'
  | 'AllocationTemplate'
  | 'AssetClass'
  | 'RebalancingThreshold';

export type AuditAction =
  | 'ASSIGN_RISK_BAND'
  | 'GENERATE_ALLOCATION_RECOMMENDATION'
  | 'GENERATE_REBALANCING_RECOMMENDATION'
  | 'ACCEPT_REBALANCING_RECOMMENDATION'
  | 'DISMISS_REBALANCING_RECOMMENDATION'
  | 'OVERRIDE_RISK_BAND'
  | 'LOG_MANUAL_RECOMMENDATION'
  | 'PUBLISH_RISK_BAND_RULE'
  | 'PUBLISH_ALLOCATION_TEMPLATE'
  | 'PUBLISH_REBALANCING_THRESHOLD'
  | 'CREATE_ASSET_CLASS'
  | 'UPDATE_ASSET_CLASS';

export type AuditDetailValue = string | number | boolean;

export interface QuestionOption {
  value: string;
  label: string;
  points: number;
}

export interface Question {
  question_id: string;
  text: string;
  options: QuestionOption[];
}

export interface Questionnaire {
  questions: Question[];
}

export interface BandRange {
  risk_band: RiskBand;
  min_points: number;
  max_points: number;
}

export interface ScoringRules {
  bands: BandRange[];
}

/** Stored form: integer basis points, summing to exactly 10000 (data-models.md §4.6). */
export interface AllocationEntry {
  asset_class_id: number;
  percent_bps: number;
}

export interface AllocationSet {
  allocations: AllocationEntry[];
}

export interface ProposedAction {
  asset_class_id: number;
  asset_class_code: string;
  action: TradeAction;
  /** Money, 2 decimals: "3200.00". */
  amount: string;
  /** Units, 4 decimals: "24.1600". */
  units: string;
  /** Signed percentage, 2 decimals: "-6.40". */
  drift_percent: string;
}

export interface ProposedActions {
  actions: ProposedAction[];
  threshold_bps: number;
  threshold_version: number;
}

export interface User {
  id: number;
  email: string;
  password_hash: string;
  role: Role;
  created_at: string;
}

export interface Customer {
  id: number;
  user_id: number;
  kyc_verified: boolean;
  created_at: string;
}

export interface RiskProfileAnswer {
  id: number;
  customer_id: number;
  question_id: string;
  answer_value: string;
  submitted_at: string;
}

export interface RiskBandAssignment {
  id: number;
  customer_id: number;
  risk_band: RiskBand;
  rule_version: number;
  assigned_at: string;
}

export interface RiskBandRule {
  id: number;
  version: number;
  questionnaire_json: Questionnaire;
  scoring_rules_json: ScoringRules;
  published_at: string;
  is_active: boolean;
}

export interface AllocationTemplate {
  id: number;
  version: number;
  risk_band: RiskBand;
  allocations_json: AllocationSet;
  published_at: string;
  is_active: boolean;
}

export interface Goal {
  id: number;
  customer_id: number;
  /** Money, 2 decimals: "250000.00". */
  target_amount: string;
  target_date: string;
  priority: number;
  created_at: string;
  updated_at: string;
}

export interface GoalProgressSnapshot {
  id: number;
  goal_id: number;
  /** Money, 2 decimals: "62500.00". */
  current_value: string;
  /** Percentage, 2 decimals, capped at "200.00". */
  percent_complete: string;
  snapshot_at: string;
  price_date: string;
}

export interface AssetClass {
  id: number;
  code: string;
  name: string;
}

export interface NavSnapshot {
  id: number;
  asset_class_id: number;
  price_date: string;
  /** Money, 2 decimals: "132.45". */
  nav_value: string;
}

export interface Holding {
  id: number;
  customer_id: number;
  asset_class_id: number;
  /** Money, 2 decimals: "48200.00". */
  current_value: string;
  as_of_date: string;
}

export interface RebalancingRecommendation {
  id: number;
  customer_id: number;
  /** Public UUID4 used in API paths. */
  recommendation_id: string;
  proposed_actions_json: ProposedActions;
  status: RecommendationStatus;
  generated_at: string;
  resolved_at: string | null;
}

export interface AdvisorOverride {
  id: number;
  customer_id: number;
  advisor_id: number;
  previous_band: RiskBand;
  new_band: RiskBand;
  reason: string;
  note: string | null;
  created_at: string;
}

export interface AuditLogEntry {
  id: number;
  entity_type: AuditEntityType;
  entity_id: string;
  actor_id: number;
  actor_role: Role;
  action: AuditAction;
  timestamp: string;
  details_json: Record<string, AuditDetailValue>;
}

export interface RebalancingThreshold {
  id: number;
  version: number;
  /** Basis points — the documented integer exception to the string rule. */
  threshold_bps: number;
  published_at: string;
  is_active: boolean;
}
